book/internals/wrap_gtk_widget.nim

```diff
@@ -0,0 +1,201 @@
+If the proc requires a parameter of type `GtkWidget` or an array/list of `GtkWidget`, add a field of type `Widget`/`seq[Widget]` to your wrapped widget instead of `GtkWidget`. 
+
+At minimum, try to wrap all features exposed by your widget directly and possibly its direct parent (e.g. for `Gtk.Scale` everything from there as well as its parent `Gtk.Range`)
+
+Follow and repeat the following steps to add a feature:
+- Wrap the GTK functions as explained in earlier sections
+- Add a field to your widget for every parameter required by the wrapped GTK functions.
+  If a field already exists because you use it during initialization with the constructor, then you don't need to add a new field.
+- Add a `property` hook for every field on your widget that is *not* of type `Widget` and that you have a wrapped GTK function for.
+  The hook should do nothing but call the wrapped GTK function with values from the implicitly available `state` variable.
+- Add a `(build, update)` hook for every field on your widget that *is* of type `Widget`.
+  The hook should do nothing but call `state.updateChild` or `state.updateChildren` with the wrapped GTK functions for adding/removing a `GtkWidget`.
+- Run your example application to check whether the added field works.
+
+
+### *Add Signal Event Listeners*
+Now we can enable the Owlkettle widget to react to GTK signals!
+Note that this section is irrelevant for a given Widget if the Widget or its parent do not provide any Signals.
+
+##### 1) Add the proc signature of signal-handler procs under the widget fields. 
+First we need to define what shall get executed when a signal gets fired. 
+What shall get executed are signal-handler procs, which are user-defined callback functions.
+
+To enable defining a signal-handler, simply define the signature of said signal-handler procs.
+These procs should never have a return-type.
+Also try to name the signal-handler proc like the GTK signal, for easier searchability in the GTK docs.
+
+For examples of signatures for signal-handler procs, go to [widgets.nim](https://github.com/can-lehmann/owlkettle/blob/main/owlkettle/widgets.nim) and look for proc signatures in between widget fields and hook definitions.
+
+##### 2) Add a `connectEvents` hook to the widget type in `widgets.nim`
+When the Owlkettle widget gets created, it needs to register its signal-handler procs with the GTK widget.
+Only then can GTK trigger their execution whenever it fires a signal corresponding to those handler procs.
+
+For this, add the `connectEvents` hook and in it, call the provided `connect` proc: `state.connect(state.<procName>, "<signalName>", <eventCallback>)`
+Where:
+  - `<procName>` is the name of the signal-handler proc
+  - `<signalName>` is the name of the GTK signal
+  - `<eventCallback>` is a proc that connects the signal-handler to the signal and updates the internal state if necessary. 
```

Comment:
```suggestion
For this, add the `connectEvents` hook and in it, call the provided `connect` proc: `state.connect(state.<procName>, "<signalName>", <eventCallback>)`
Where:
  - `<eventName>` is the name of the Owlkettle event
  - `<signalName>` is the name of the GTK signal
  - `<eventCallback>` is the the signal-handler
```

Reply:
Is `state.<procName>` not to get a pointer or sth. of the proc that will be called in eventCallback?

Reply:
Yes, the thing that is called in `eventCallback` is the Owlkettle event. The main point of the suggestion was to get rid of the word `signal-handler` to describe `procName`, as `eventCallback` is the signal-handler (= the thing that is called directly be GTK).

The definition of connect is

```nim
proc connect*[T](renderable: Renderable,
                 event: Event[T],
                 name: cstring,
                 eventCallback: pointer) =
```

so technically `event` is not actually a proc.


I fixed the signature in the suggestion.

Reply:
Check, that one makes sense to me. 
I adjusted the wording in the line before this paragraph a bit to account for this as that makes it a bit clearer (I hope) and fixed a typo in the suggestion (that's why I didn't commit the suggestion directly):

```txt
For this, add the `connectEvents` hook and in it, call the provided `connect` proc: `state.connect(state.<eventName>, "<signalName>", <eventCallback>)`
Where:
  - `<eventName>` is the name of the Owlkettle event
  - `<signalName>` is the name of the GTK signal
  - `<eventCallback>` is the signal-handler
  ```